### DISCUSSION TITLE
Program requests/fulfillment

### DISCUSSION THREAD

**Discussion by justinfrevert** [justinfrevert/Project-TOEP#2] [General]
Context: as of now, the programs are stored onchain, as well as the proofs.

There should be some way for users to request programs be developed and uploaded by the community. However, without allowing for criteria for onchain verification, I am not sure what that would look like - if we are able to build it now, then we should do it. One idea is to just have a field for a text description as a placeholder where the user would describe the desired functionality. 

There should be a program upload functionality that specifically responds to someone's request for a program. The user uploading the program should receive some small static reward amount.

**Comment by vivekvpandya** [comment]
I think here payment for developing program becomes bit tricky.
I feel it needs some sort of escrow because requestor should only be charged if his requirements are mostly ful-filled. And developer should not be denied to get paid for certain long duration if he has correctly done the job.

I think better to start with chain storing ELF and porgram_id and provers can generate proof by running Risc0 on their HW.
On chain logic can check if proof verifies for given program_id it just pays from locked/reserved funds of requester.

**Reply by justinfrevert** [reply]
Another thought: there are already plenty of pallets that are very specialized towards community work and rewards such as bounties and treasury. Maybe checks on the program id, as well as the requester lock/reserve funds can be integrated with those.

Overall, I agree, including the point about lock/reserving the requester's funds

**Comment by vivekvpandya** [comment]
if we require users to upload their Rust programs or ask other to write their Rust programs it may not be very attractive to users who wish to have some private code proven. 
In simple proof market configuration we don't need offchain service too, A requester just post (ELF, image_id and reward ) on market and other users uploads proof , chain checks if it verifies uploaded of proof gets reward.

**Comment by justinfrevert** [comment]
@vivekvpandya we should probably remove program storage then, since `image_id` is all that is needed. I can add an issue for it

**Comment by justinfrevert** [comment]
Actually, I remembered in the current design, the Prover gets the bytes of each program from onchain. Not sure how to ensure data availability of programs, while preserving privacy of their contents. Perhaps some private messaging from developer to prover who is set to prove some program's execution?
